internal/bootstrap/deployer.go

```diff
@@ -0,0 +1,498 @@
+	if err != nil {
+		return nil, nil, errors.Annotatef(err, "cannot store controller charm at %q", controllerCharmPath)
+	}
+	b.logger.Debugf("Successfully deployed local Juju controller charm")
+	origin := corecharm.Origin{
+		Source: corecharm.Local,
+		Type:   "charm",
+		Platform: corecharm.Platform{
+			Architecture: arch,
+			OS:           base.OS,
+			Channel:      base.Channel.String(),
+		},
+	}
+	return curl, &origin, nil
+}
+
+// DeployCharmhubCharm deploys the controller charm from charm hub.
+func (b *baseDeployer) DeployCharmhubCharm(ctx context.Context, arch string, base corebase.Base) (string, *corecharm.Origin, error) {
+	model, err := b.stateBackend.Model()
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+
+	charmRepo, err := b.newCharmRepo(services.CharmRepoFactoryConfig{
+		LoggerFactory:      b.loggerFactory,
+		CharmhubHTTPClient: b.charmhubHTTPClient,
+		StateBackend:       b.charmUploader,
+		ModelBackend:       model,
+	})
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+
+	var curl *charm.URL
+	if b.charmhubURL == "" {
+		curl = charm.MustParseURL(controllerCharmURL)
+	} else {
+		curl = charm.MustParseURL(b.charmhubURL)
+	}
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+	origin := corecharm.Origin{
+		Source:  corecharm.CharmHub,
+		Type:    "charm",
+		Channel: &b.channel,
+		Platform: corecharm.Platform{
+			Architecture: arch,
+			OS:           base.OS,
+			Channel:      base.Channel.Track,
+		},
+	}
+
+	// Since we're running on the machine to which the controller charm will be
+	// deployed, we know the exact platform to ask for, not need to review the
+	// supported series.
+	//
+	// We prefer the latest LTS series, if the current series is not one,
+	// charmRepo.ResolveWithPreferredChannel, will return an origin with the
+	// latest LTS based on data provided by charmhub in the revision-not-found
+	// error response.
+	//
+	// The controller charm doesn't have any series specific code.
+	curl, origin, _, err = charmRepo.ResolveWithPreferredChannel(ctx, curl.Name, origin)
+	if err != nil {
+		return "", nil, errors.Annotatef(err, "resolving %q", controllerCharmURL)
+	}
+
+	charmDownloader, err := b.newCharmDownloader(services.CharmDownloaderConfig{
+		LoggerFactory:      b.loggerFactory,
+		CharmhubHTTPClient: b.charmhubHTTPClient,
+		ObjectStore:        b.objectStore,
+		StateBackend:       b.charmUploader,
+		ModelBackend:       model,
+	})
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+	resOrigin, err := charmDownloader.DownloadAndStore(ctx, curl, origin, false)
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+
+	b.logger.Debugf("Successfully deployed charmhub Juju controller charm")
+
+	return curl.String(), &resOrigin, nil
+}
+
+// AddControllerApplication adds the controller application.
+func (b *baseDeployer) AddControllerApplication(ctx context.Context, curl string, origin corecharm.Origin, controllerAddress string) (Unit, error) {
```

Comment:
Is this context going to be used in future? It isn't at the moment. There are other unused ones in _deployer_iaas.go_ to.

Reply:
Yes, context is everywhere.